mpl_gui/__init__.py

```diff
@@ -198,7 +232,7 @@ def show_all(self, *, block=None, timeout=None):
 
         if timeout is None:
             timeout = self._timeout
-
+        self._promote_and_number()
```

Comment:
Why are figures only promoted on show_all?  Maybe have a comment explaining this?  I'm also not clear on the "promoted" terminology or why you need this concept.   It would seem to me that all figures would just be added to the registry and numbered as they come in...

Reply:
How we got to this weird state:

 - the goal is to create `Figure` objects without immediately creating the `FigureManger` + `FigureCanvas` because the later two are backend-dependent and in most (all?) cases create objects in the c/c++ layer
 - the notion of "number" lives on `FigureManagerBase` not in `Figure`
 - thus deferred numbering the Figures until we are forced to actually create the `FigureManager`

It would be reasonable to number the `Figures` when they are created, but I was being lazy about creating extra state...

Reply:
and right after writing this I went back to read the code and they are already getting numbered on creation!  re-working this...

Reply:
And now I remember the other complexity I was avoiding.

The counting we are doing at figure registration time is montonic figure count (because that is simple!) but was trying to reproduce the "restarting at 1" behavior of pyplot.  I think the options here are:
 - abandon the pyplot-like numbering and just accept manager monotonic increasing 
 - abandon the monotonic counting
In either case we need some way to associate the number at `Figure` creation time with the `Figure`.  Options I can think of:
 - start to keep a weak-key dict mapping `Figure` -> number (I tried very hard to make there only be one place in `FigureRegistry` that we keep a hard ref to the `Figure` (in `FigureRegistry.figures`, all of the other views are strictly derived from this).  While we can not keep users from keeping hard-refs, we should be as sure as possible that we do not leak them
 - monkey patch `Figure` to carry the number

I do not think we should push the notion "number" into `Figure` itself as there is an implication that the number is (temporarily) unique within a process/registry and part of the goal here is to pull apart `Figure`,  `FigureManager`, and the `Registry`. 

I'm leaning towards monkey-patching for now, but changing the canonical version of the figure refernce to be a dictionary mapping int -> `Figure` also seems like a good option.